### PR TITLE
[TextServer] Implement soft hyphen handling.

### PR DESCRIPTION
**Pull request by bruvzg**
Fixes https://github.com/godotengine/godot/issues/85815

https://github.com/godotengine/godot/assets/7645683/b17efe29-a864-4603-83e9-17c12098dc64

### CODE REVIEW

File: modules/text_server_adv/text_server_adv.cpp [hunk 1]
```diff
@@ -4526,6 +4526,12 @@ bool TextServerAdvanced::_shape_substr(ShapedTextDataAdvanced *p_new_sd, const S
 					if ((sd_glyphs[j].start >= bidi_run_start) && (sd_glyphs[j].end <= bidi_run_end)) {
 						// Copy glyphs.
 						Glyph gl = sd_glyphs[j];
+						if (gl.end == p_start + p_length && ((gl.flags & GRAPHEME_IS_SHY) == GRAPHEME_IS_SHY)) {
+							uint32_t index = font_get_glyph_index(gl.font_rid, gl.font_size, 0x00ad, 0);
+							float w = (p_new_sd->orientation == ORIENTATION_HORIZONTAL) ? font_get_glyph_advance(gl.font_rid, gl.font_size, index).x : font_get_glyph_advance(gl.font_rid, gl.font_size, index).y;
```
Comment (AThousandShips) [inline modules/text_server_adv/text_server_adv.cpp:4531]:
```suggestion
							float w = 
 font_get_glyph_advance(gl.font_rid, gl.font_size, index)[(p_new_sd->orientation == ORIENTATION_HORIZONTAL) ? 0 : 1];
```
Alternative, compact approach, also more readable IMO as you don't have to ensure both sides of the ternary are the same

File: servers/text_server.cpp [hunk 2]
```diff
@@ -823,8 +825,17 @@ PackedInt32Array TextServer::shaped_text_get_line_breaks_adv(const RID &p_shaped
 			}
 			if (p_break_flags.has_flag(BREAK_WORD_BOUND)) {
 				if ((l_gl[i].flags & GRAPHEME_IS_BREAK_SOFT) == GRAPHEME_IS_BREAK_SOFT) {
-					last_safe_break = i;
-					word_count++;
+					if ((l_gl[i].flags & GRAPHEME_IS_SHY) == GRAPHEME_IS_SHY) {
+						uint32_t gl = font_get_glyph_index(l_gl[i].font_rid, l_gl[i].font_size, 0x00ad, 0);
+						float w = (orientation == ORIENTATION_HORIZONTAL) ? font_get_glyph_advance(l_gl[i].font_rid, l_gl[i].font_size, gl).x : font_get_glyph_advance(l_gl[i].font_rid, l_gl[i].font_size, gl).y;
```
Comment (AThousandShips) [inline servers/text_server.cpp:830]:
```suggestion
						float w = font_get_glyph_advance(l_gl[i].font_rid, l_gl[i].font_size, gl)[(orientation == ORIENTATION_HORIZONTAL) ? 0 : 1];
```

File: servers/text_server.cpp [hunk 3]
```diff
@@ -943,8 +955,17 @@ PackedInt32Array TextServer::shaped_text_get_line_breaks(const RID &p_shaped, do
 			}
 			if (p_break_flags.has_flag(BREAK_WORD_BOUND)) {
 				if ((l_gl[i].flags & GRAPHEME_IS_BREAK_SOFT) == GRAPHEME_IS_BREAK_SOFT) {
-					last_safe_break = i;
-					word_count++;
+					if ((l_gl[i].flags & GRAPHEME_IS_SHY) == GRAPHEME_IS_SHY) {
+						uint32_t gl = font_get_glyph_index(l_gl[i].font_rid, l_gl[i].font_size, 0x00AD, 0);
+						float w = (orientation == ORIENTATION_HORIZONTAL) ? font_get_glyph_advance(l_gl[i].font_rid, l_gl[i].font_size, gl).x : font_get_glyph_advance(l_gl[i].font_rid, l_gl[i].font_size, gl).y;
```
Comment (AThousandShips) [inline servers/text_server.cpp:960]:
```suggestion
						float w = font_get_glyph_advance(l_gl[i].font_rid, l_gl[i].font_size, gl)[(orientation == ORIENTATION_HORIZONTAL) ? 0 : 1];
```

File: servers/text_server.h [hunk 4]
```diff
@@ -141,6 +141,7 @@ class TextServer : public RefCounted {
 		GRAPHEME_IS_CONNECTED = 1 << 10, // Connected to previous grapheme.
 		GRAPHEME_IS_SAFE_TO_INSERT_TATWEEL = 1 << 11, // It is safe to insert a U+0640 before this grapheme for elongation.
 		GRAPHEME_IS_EMBEDDED_OBJECT = 1 << 12, // Grapheme is an object replacement character for the embedded object.
+		GRAPHEME_IS_SHY = 1 << 13, // Grapheme is a soft hyphen.
```
Comment (akien-mga) [inline servers/text_server.h:144]: I assume `SHY` might be the Unicode name for the character, but would it make sense to spell it out as `GRAPHEME_IS_SOFT_HYPHEN` to be more explicit and not sound like the grapheme is socially conscious? :P

**Review by akien-mga — APPROVED** [review]
Looks good to me.

File: servers/text_server.h (hunk 4 above)
Comment (bruvzg) [reply servers/text_server.h]: Renamed to `GRAPHEME_IS_SOFT_HYPHEN`.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**